### PR TITLE
Add functionality for a custom outer boundary condition on `br`

### PR DESCRIPTION
**Pull request by TrestanSimon**
## PR Description

This is a continuation of https://github.com/dstansby/pfsspy/pull/357 and https://github.com/dstansby/pfsspy/pull/362 which were closed due to inactivity and pfsspy being archived, respectively. From the description of the original PR:

> This PR adds the option to solve using an arbitrary map for the outer boundary condition on `br`, a functionality which is already present in Anthony Yeates's [original code](https://github.com/antyeates1983/pfss/blob/a8e1db39b8712a616ad5d11c6b92b8605bee02ae/pfss.py#L110-L115). Here, it is implemented in a similar manner. (For mathematical details, see the text following Equation 24 in the ["Notes on the Numerical Methods" document](https://github.com/dstansby/pfsspy/blob/b22eb10c47a98e44bd9b90be101a9b5c0929e269/doc/source/manual/numerical_methods.pdf).)

This PR also adds an example and test of this new functionality.

### CODE REVIEW

File: examples/using_sunkit_magex_pfss/dipole_source_solution.py [hunk 1]
```diff
@@ -123,3 +123,39 @@ def dipole_Br(r, s):
 ax.set_title('PFSS solution for a dipole source field')
 
 plt.show()
+
+###############################################################################
+# We are also able to provide a custom outer boundary condition to our
+# `sunkit_magex.pfss.Input` object. For example, we can set the radial
+# component of the magnetic field equal to zero across (outer) source surface.
```
Comment (nabobalis) [inline examples/using_sunkit_magex_pfss/dipole_source_solution.py:130]:
```suggestion
# component of the magnetic field equal to zero across the (outer) source surface.
```
?

File: examples/using_sunkit_magex_pfss/dipole_source_solution.py [hunk 2]
```diff
@@ -123,3 +123,39 @@ def dipole_Br(r, s):
 ax.set_title('PFSS solution for a dipole source field')
 
 plt.show()
```
Comment (nabobalis) [inline examples/using_sunkit_magex_pfss/dipole_source_solution.py:125]:
```suggestion
```

File: sunkit_magex/pfss/input.py [hunk 3]
```diff
@@ -23,26 +23,39 @@ class Input:
         on.
     rss : float
         Radius of the source surface, as a fraction of the solar radius.
+    br_outer : sunpy.map.GenericMap, String
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply sunkit_magex/pfss/input.py:26]: Stupid question, what kind of Map would need to be entered?

[the remark this answers is not in the document]
Comment (TrestanSimon) [reply sunkit_magex/pfss/input.py]: It should just be a GenericMap containing data with a CEA projection and matching dimensions with `br`. I just added checks for the former in 107ec35.

File: sunkit_magex/pfss/tests/test_pfss.py [hunk 4]
```diff
@@ -39,6 +39,31 @@ def test_pfss(gong_map):
     return br
 
 
+@pytest.mark.array_compare(reference_dir="arraydiff_reference")
+def test_pfss_br_outer(gong_map):
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply sunkit_magex/pfss/tests/test_pfss.py:43]: Great question, I'm going to say no. Maybe an extension for a future PR

[the remark this answers is not in the document]
Comment (STBadman) [reply sunkit_magex/pfss/tests/test_pfss.py:43]: This is a great point - there is currently a dipole result in `sunkit_magex/pfss/tests/conftest.py` - i.e. the output of pfss with radial outer boundary given an input map which is an analytical dipole, and this has a number of symmetries for the radial boundary condition case which are tested for in `sunkit_magex/pfss/tests/test_pfss` e.g. in `test_bvec_interpolator` and `test_field_line_polarity` . 

I think the way to do this would be to come up with a similar simple test result(s) with a specified outer boundary (created in conftest) whose symmetries are known and write an analogous test function. I don't immediately have a fully thought out example, but maybe just another dipole with 1/r^3 weaker strength at the source surface would work.

(Longer term, there is a set of analytical functions for the radial source surface case for a given multipole moment `sunkit_magex/pfss/analytic.py`. David Stansby has an interesting preprint comparing them (https://arxiv.org/pdf/2201.07783) which shows they're good for not perfect at least for field line tracing on a discretized grid. It would be nice to leverage these to do a more systematic test for the radial boundary condition. If there's another outer boundary that would give closed form expresseions for c_l and d_l, then tests with this new functionality could follow)

**Review by STBadman — CHANGES_REQUESTED** [review]
This is looking good to me - let's just try and get a test with a dipole lower boundary and simple (but not phi=0) outer boundary as discussed above. 

On the CI test failure I'm not sure what the fix (or problem!) is - @Cadair or @nabobalis could you advise?

**Comment by nabobalis** [comment]
The error is that we dont have the text file with the expected values for the test to compare to.

**Comment by TrestanSimon** [comment]
> The error is that we dont have the text file with the expected values for the test to compare to. That needs to be added to the PR.

The text file with the expected values for `test_pfss_br_outer` is the same as that for `test_pfss`; it is in `arraydiff_reference/test_pfss.py`. Specifying this filename in the array_compare decorator (27a52c6) seems to have worked. This was not an issue for me locally.

File: examples/using_sunkit_magex_pfss/dipole_source_solution.py [hunk 5]
```diff
@@ -122,4 +122,39 @@ def dipole_Br(r, s):
...
+pfss_in_closed = pfss.Input(input_map, nrho, rss, map_zeros)
+pfss_out_closed = pfss.pfss(pfss_in_closed)
+
+# Now trace field lines from the same starting positions as before but using
+# our new solution.
+fig, ax = plt.subplots()
+ax.set_aspect('equal')
+
+seeds = SkyCoord(lon, lat, r, frame=pfss_out_closed.coordinate_frame)
+
```
Comment (nabobalis) [inline examples/using_sunkit_magex_pfss/dipole_source_solution.py:144]:
```suggestion
```

File: sunkit_magex/pfss/input.py [hunk 6]
```diff
@@ -23,14 +23,16 @@ class Input:
         on.
     rss : float
         Radius of the source surface, as a fraction of the solar radius.
+    br_outer : sunpy.map.GenericMap, str
+        Boundary condition of radial magnetic field at the outer surface.
```
Comment (nabobalis) [inline sunkit_magex/pfss/input.py:27]:
```suggestion
    br_outer : sunpy.map.GenericMap, str, optional
        Boundary condition of radial magnetic field at the outer surface.
        Defaults to "radial".
```

File: sunkit_magex/pfss/input.py [hunk 7]
```diff
@@ -23,8 +23,9 @@ class Input:
         on.
     rss : float
         Radius of the source surface, as a fraction of the solar radius.
-    br_outer : sunpy.map.GenericMap, str
+    br_outer : sunpy.map.GenericMap, str, optional
```
Comment (STBadman) [inline sunkit_magex/pfss/input.py:26]: i think this is tripping the docs build, the syntax might need to be 

`sunpy.map.GenericMap or str, optional`

or 

`Optional[sunpy.map.GenericMap or str]` 

are you able to test the docs build locally?

Comment (nabobalis) [reply sunkit_magex/pfss/input.py:26]: Oh yeah, my suggestion was wrong.

Comment (TrestanSimon) [reply sunkit_magex/pfss/input.py]: Yes, I can test the docs build locally. I have accidentally been doing so on a separate, fresh download of sunkit-magex, so I have been missing these errors. Hence why I pushed 98b2105. I did not notice this until now. My apologies.

**Review by STBadman — APPROVED** [review]
I think this looks good to go. Thank you for pursuing this @TrestanSimon !

**Comment by nabobalis** [comment]
Thank you @TrestanSimon and @STBadman
